File: backend/app/schemas/ulpin.py
import uuid
from typing import Optional
from pydantic import BaseModel, Field


class ULPIN3DGenerateRequest(BaseModel):
    strata_unit_id: uuid.UUID
    base_ulpin: str = Field(..., min_length=14, max_length=14, description="14-char standard Bhu-Aadhaar")
    tower_number: str
    floor_number: int
    unit_number: str
    base_elevation_amsl: float


class ULPINVerifyResponse(BaseModel):
    is_valid: bool
    ulpin_3d: str
    base_ulpin: str
    verification_code: str
    status: str
    spatial_hash: str
    qr_payload: str

File: frontend/src/types/ulpin.ts
export interface ULPIN3DRecord {
  id: string;
  strata_unit_id: string;
  base_ulpin: string;
  ulpin_3d: string;
  spatial_hash: string;
  verification_code: string;
  qr_payload: string;
  status: "GENERATED" | "ISSUED" | "REVOKED";
  encoding_version: string;
}

export interface ULPINGenerationInput {
  strata_unit_id: string;
  base_ulpin: string;
  tower_number: string;
  floor_number: number;
  unit_number: string;
  base_elevation_amsl: number;
}
